README
NAME
    Tags::HTML::Image::Grid - Tags helper for image grid.

SYNOPSIS
     use Tags::HTML::Image::Grid;

     my $obj = Tags::HTML::Image::Grid->new(%params);
     $obj->process($images_ar);
     $obj->process_css;

METHODS
  "new"
     my $obj = Tags::HTML::Image::Grid->new(%params);

    Constructor.

    *       "css"

            'CSS::Struct::Output' object for process_css processing.

            Default value is undef.

    *       "css_image_grid"

            Form CSS style.

            Default value is 'image-grid'.

    *       "img_border_color_cb"

            Image border callback. This is used with 'img_border_width'
            parameter. Border will be present only for image which has color
            defined.

            Callback arguments are: $self and $image object.

            Default value is undef.

    *       "img_border_width"

            Image border width if need to use.

            Default value is undef.

    *       "img_link_cb"

            Image link callback.

            Default value is undef.

    *       "img_select_cb"

            Image select callback.

            Default value is undef.

    *       "img_src_cb"

            Image src callback across data object.

            Default value is undef.

    *       "img_width"

            Image width in pixels.

            Default value is 340.

    *       "tags"

            'Tags::Output' object.

            Default value is undef.

    *       "title"

            Image grid title.

            Default value is undef.

  "process"
     $obj->process($images_ar);

    Process Tags structure for images in $images_ar to output.

    Accepted items in $images_ar reference to array are Data::Image objects.

    Returns undef.

  "process_css"
     $obj->process_css;

    Process CSS::Struct structure for output.

    Returns undef.

ERRORS
     new():
             From Class::Utils::set_params():
                     Unknown parameter '%s'.
             From Tags::HTML::new():
                     Parameter 'css' must be a 'CSS::Struct::Output::*' class.
                     Parameter 'tags' must be a 'Tags::Output::*' class.
             Parameter 'img_link_cb' must be a code.
             Parameter 'img_select_cb' must be a code.
             Parameter 'img_src_cb' must be a code.

     process():
             From Tags::HTML::process():
                     Parameter 'tags' isn't defined.
             Bad data image object.

     process_css():
             From Tags::HTML::process_css():
                     Parameter 'css' isn't defined.

EXAMPLE
     use strict;
     use warnings;

     use CSS::Struct::Output::Indent;
     use Data::Image;
     use Tags::HTML::Image::Grid;
     use Tags::Output::Indent;

     # Object.
     my $css = CSS::Struct::Output::Indent->new;
     my $tags = Tags::Output::Indent->new;
     my $obj = Tags::HTML::Image::Grid->new(
             'css' => $css,
             'tags' => $tags,
     );

     # Images.
     my $image1 = Data::Image->new(
             'comment' => 'Michal from Czechia',
             'url' => 'https://upload.wikimedia.org/wikipedia/commons/a/a4/Michal_from_Czechia.jpg',
     );
     my $image2 = Data::Image->new(
             'comment' => 'Self photo',
             'url' => 'https://upload.wikimedia.org/wikipedia/commons/7/76/Michal_Josef_%C5%A0pa%C4%8Dek_-_self_photo_3.jpg',
     );

     # Process image grid.
     $obj->process([$image1, $image2]);
     $obj->process_css;

     # Print out.
     print $tags->flush;
     print "\n\n";
     print $css->flush;

     # Output:
     # <div class="image-grid">
     #   <div class="image-grid-inner">
     #     <figure>
     #       <img src=
     #         "https://upload.wikimedia.org/wikipedia/commons/a/a4/Michal_from_Czechia.jpg"
     #         >
     #       </img>
     #       <figcaption>
     #         Michal from Czechia
     #       </figcaption>
     #     </figure>
     #     <figure>
     #       <img src=
     #         "https://upload.wikimedia.org/wikipedia/commons/7/76/Michal_Josef_%C5%A0pa%C4%8Dek_-_self_photo_3.jpg"
     #         >
     #       </img>
     #       <figcaption>
     #         Self photo
     #       </figcaption>
     #     </figure>
     #   </div>
     # </div>
     # 
     # .image-grid {
     #      display: flex;
     #      align-items: center;
     #      justify-content: center;
     # }
     # .image-grid-inner {
     #      display: grid;
     #      grid-gap: 1px;
     #      grid-template-columns: repeat(4, 340px);
     # }
     # .image-grid figure {
     #      object-fit: cover;
     #      width: 340px;
     #      height: 340px;
     #      position: relative;
     #      overflow: hidden;
     #      border: 1px solid white;
     #      margin: 0;
     #      padding: 0;
     # }
     # .image-grid img {
     #      object-fit: cover;
     #      width: 100%;
     #      height: 100%;
     #      vertical-align: middle;
     # }
     # .image-grid figcaption {
     #      margin: 0;
     #      padding: 1em;
     #      position: absolute;
     #      z-index: 1;
     #      bottom: 0;
     #      left: 0;
     #      width: 100%;
     #      max-height: 100%;
     #      overflow: auto;
     #      box-sizing: border-box;
     #      transition: transform 0.5s;
     #      transform: translateY(100%);
     #      background: rgba(0, 0, 0, 0.7);
     #      color: rgb(255, 255, 255);
     # }
     # .image-grid figure:hover figcaption {
     #      transform: translateY(0%);
     # }
     # .image-grid .selected {
     #      border: 1px solid black;
     #      border-radius: 0.5em;
     #      color: black;
     #      padding: 0.5em;
     #      position: absolute;
     #      right: 10px;
     #      top: 10px;
     # }

DEPENDENCIES
    Class::Utils, Error::Pure, List::MoreUtils, Scalar::Util, Tags::HTML,
    Unicode::UTF8.

REPOSITORY
    <https://github.com/michal-josef-spacek/Tags-HTML-Image-Grid>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2022 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.04



Changes
0.04

0.03 2022-12-04T18:03:23+01:00
 - Add parameters (img_border_color_cb and img_border_width) to use border
   for some images. img_border_width is global parameter and color could be
   variable.
 - Fix border of image.
 - Fix existing of value in comment.

0.02 2022-11-05T17:33:05+01:00
 - Fix existing of value in select callback.

0.01 2022-10-14T12:58:51+02:00
 - First version.


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Tags helper class for image grid.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.03;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Tags-HTML-Image-Grid';
readme_from 'Grid.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.02;
requires 'Error::Pure' => 0.07;
requires 'List::MoreUtils' => 0;
requires 'Unicode::UTF8' => 0;
requires 'Tags::HTML' => 0;
requires 'perl' => '5.8.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Tags-HTML-Image-Grid/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Tags-HTML-Image-Grid';
resources 'repository' => 'git://github.com/michal-josef-spacek/Tags-HTML-Image-Grid';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'File::Object' => 0.03;
test_requires 'Tags::Output::Raw' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.04';

# Run.
WriteAll();


Grid.pm
package Tags::HTML::Image::Grid;

use base qw(Tags::HTML);
use strict;
use warnings;

use Class::Utils qw(set_params split_params);
use Error::Pure qw(err);
use List::MoreUtils qw(none);
use Scalar::Util qw(blessed);
use Unicode::UTF8 qw(decode_utf8);

our $VERSION = 0.04;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my ($object_params_ar, $other_params_ar) = split_params(
		['css_image_grid', 'img_border_color_cb', 'img_border_width', 'img_link_cb',
		'img_select_cb', 'img_src_cb', 'img_width', 'title'], @params);
	my $self = $class->SUPER::new(@{$other_params_ar});

	# Form CSS style.
	$self->{'css_image_grid'} = 'image-grid';

	# Image border color callback.
	$self->{'img_border_color_cb'} = undef;

	# Image border width (in pixels).
	$self->{'img_border_width'} = undef;

	# Image link callback.
	$self->{'img_link_cb'} = undef;

	# Image select callback.
	$self->{'img_select_cb'} = undef;

	# Image src callback across data object.
	$self->{'img_src_cb'} = undef;

	# Image width in pixels.
	$self->{'img_width'} = 340;

	# Image grid title.
	$self->{'title'} = undef;

	# Process params.
	set_params($self, @{$object_params_ar});

	# Check callback codes.
	$self->_check_callback('img_border_color_cb');
	$self->_check_callback('img_link_cb');
	$self->_check_callback('img_select_cb');
	$self->_check_callback('img_src_cb');

	# Object.
	return $self;
}

sub _check_callback {
	my ($self, $callback_key) = @_;

	if (defined $self->{$callback_key}
		&& ref $self->{$callback_key} ne 'CODE') {

		err "Parameter '$callback_key' must be a code.";
	}

	return;
}

sub _check_images {
	my ($self, $images_ar) = @_;

	foreach my $image (@{$images_ar}) {
		if (! blessed($image) && ! $image->isa('Data::Image')) {

			err 'Bad data image object.';
		}
	}

	return;
}

# Process 'Tags'.
sub _process {
	my ($self, $images_ar) = @_;

	$self->_check_images($images_ar);

	$self->{'tags'}->put(
		['b', 'div'],
		['a', 'class', $self->{'css_image_grid'}],

		['b', 'div'],
		['a', 'class', $self->{'css_image_grid'}.'-inner'],
	);
	if (defined $self->{'title'}) {
		$self->{'tags'}->put(
			['b', 'fieldset'],
			['b', 'legend'],
			['d', $self->{'title'}],
			['e', 'legend'],
		);
	}
	foreach my $image (@{$images_ar}) {

		# Begin of image link.
		if (defined $self->{'img_link_cb'}) {
			$self->{'tags'}->put(
				['b', 'a'],
				['a', 'href', $self->{'img_link_cb'}->($image)],
			);
		}

		# Begin of figure.
		$self->{'tags'}->put(
			['b', 'figure'],
		);

		# Select information.
		if (defined $self->{'img_select_cb'}) {
			my $select_hr = $self->{'img_select_cb'}->($self, $image);
			if (ref $select_hr eq 'HASH' && exists $select_hr->{'value'}) {
				$select_hr->{'css_background_color'} ||= 'lightgreen';
				$self->{'tags'}->put(
					['b', 'i'],
					['a', 'class', 'selected'],
					['a', 'style', 'background-color: '.$select_hr->{'css_background_color'}.';'],
					exists $select_hr->{'value'} ? (
						['d', $select_hr->{'value'}],
					) : (),
					['e', 'i'],
				);
			}
		}

		# Image.
		my $image_url;
		if (defined $image->url) {
			$image_url = $image->url;
		} elsif (defined $image->url_cb) {
			$image_url = $image->url_cb->($image);
		} elsif (defined $self->{'img_src_cb'}) {
			$image_url = $self->{'img_src_cb'}->($image);
		} else {
			err 'No image URL.';
		}
		my $img_style;
		if (defined $self->{'img_border_width'} && defined $self->{'img_border_color_cb'}) {
			my $border_color = $self->{'img_border_color_cb'}->($self, $image);
			if (defined $border_color) {
				$img_style = 'border:'.$self->{'img_border_width'}.'px solid '.$border_color.';';
			}
		}
		$self->{'tags'}->put(
			['b', 'img'],
			defined $img_style ? (
				['a', 'style', $img_style],
				['a', 'class', 'border'],
			) : (),
			['a', 'src', $image_url],
			['e', 'img'],
		);

		# Image comment.
		if (defined $image->comment) {
			$self->{'tags'}->put(
				['b', 'figcaption'],
				['d', $image->comment],
				['e', 'figcaption'],
			);
		}

		# Enf of figure.
		$self->{'tags'}->put(
			['e', 'figure'],
		);

		# End of image link.
		if (defined $self->{'img_link_cb'}) {
			$self->{'tags'}->put(
				['e', 'a'],
			);
		}
	}
	if (defined $self->{'title'}) {
		$self->{'tags'}->put(
			['e', 'fieldset'],
		);
	}
	$self->{'tags'}->put(
		['e', 'div'],
		['e', 'div'],
	);

	return;
}

sub _process_css {
	my $self = shift;

	# Compute image width and height.
	my ($img_width, $img_height) = ($self->{'img_width'}, $self->{'img_width'});
	if (defined $self->{'img_border_width'}) {
		$img_width -= 2 * $self->{'img_border_width'};
		$img_height -= 2 * $self->{'img_border_width'};
	}

	$self->{'css'}->put(

		# Grid center on page.
		['s', '.'.$self->{'css_image_grid'}],
		['d', 'display', 'flex'],
		['d', 'align-items', 'center'],
		['d', 'justify-content', 'center'],
		['e'],

		# 4 columns in grid.
		['s', '.'.$self->{'css_image_grid'}.'-inner'],
		['d', 'display', 'grid'],
		['d', 'grid-gap', '1px'],
		['d', 'grid-template-columns', 'repeat(4, '.$self->{'img_width'}.'px)'],
		['e'],

		# Create rectangle.
		['s', '.'.$self->{'css_image_grid'}.' figure'],
		['d', 'object-fit', 'cover'],
		['d', 'width', $self->{'img_width'}.'px'],
		['d', 'height', $self->{'img_width'}.'px'],
		['d', 'position', 'relative'],
		['d', 'overflow', 'hidden'],
		['d', 'margin', 0],
		['d', 'padding', 0],
		['e'],

		['s', '.'.$self->{'css_image_grid'}.' img'],
		['d', 'object-fit', 'cover'],
		['d', 'width', '100%'],
		['d', 'height', '100%'],
		['d', 'vertical-align', 'middle'],
		['e'],

		['s', '.'.$self->{'css_image_grid'}.' img.border'],
		['d', 'object-fit', 'cover'],
		['d', 'width', $img_width.'px'],
		['d', 'height', $img_height.'px'],
		['d', 'vertical-align', 'middle'],
		['e'],

		['s', '.'.$self->{'css_image_grid'}.' figcaption'],
		['d', 'margin', 0],
		['d', 'padding', '1em'],
		['d', 'position', 'absolute'],
		['d', 'z-index', 1],
		['d', 'bottom', 0],
		['d', 'left', 0],
		['d', 'width', '100%'],
		['d', 'max-height', '100%'],
		['d', 'overflow', 'auto'],
		['d', 'box-sizing', 'border-box'],
		['d', 'transition', 'transform 0.5s'],
		['d', 'transform', 'translateY(100%)'],
		['d', 'background', 'rgba(0, 0, 0, 0.7)'],
		['d', 'color', 'rgb(255, 255, 255)'],
		['e'],

		['s', '.'.$self->{'css_image_grid'}.' figure:hover figcaption'],
		['d', 'transform', 'translateY(0%)'],
		['e'],

		['s', '.'.$self->{'css_image_grid'}.' .selected'],
		['d', 'border', '1px solid black'],
		['d', 'border-radius', '0.5em'],
		['d', 'color', 'black'],
		['d', 'padding', '0.5em'],
		['d', 'position', 'absolute'],
		['d', 'right', '10px'],
		['d', 'top', '10px'],
		['e'],
	);

	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Tags::HTML::Image::Grid - Tags helper for image grid.

=head1 SYNOPSIS

 use Tags::HTML::Image::Grid;

 my $obj = Tags::HTML::Image::Grid->new(%params);
 $obj->process($images_ar);
 $obj->process_css;

=head1 METHODS

=head2 C<new>

 my $obj = Tags::HTML::Image::Grid->new(%params);

Constructor.

=over 8

=item * C<css>

'CSS::Struct::Output' object for L<process_css> processing.

Default value is undef.

=item * C<css_image_grid>

Form CSS style.

Default value is 'image-grid'.

=item * C<img_border_color_cb>

Image border callback. This is used with 'img_border_width' parameter.
Border will be present only for image which has color defined.

Callback arguments are: C<$self> and C<$image> object.

Default value is undef.

=item * C<img_border_width>

Image border width if need to use.

Default value is undef.

=item * C<img_link_cb>

Image link callback.

Default value is undef.

=item * C<img_select_cb>

Image select callback.

Default value is undef.

=item * C<img_src_cb>

Image src callback across data object.

Default value is undef.

=item * C<img_width>

Image width in pixels.

Default value is 340.

=item * C<tags>

'Tags::Output' object.

Default value is undef.

=item * C<title>

Image grid title.

Default value is undef.

=back

=head2 C<process>

 $obj->process($images_ar);

Process Tags structure for images in C<$images_ar> to output.

Accepted items in C<$images_ar> reference to array are L<Data::Image> objects.

Returns undef.

=head2 C<process_css>

 $obj->process_css;

Process CSS::Struct structure for output.

Returns undef.

=head1 ERRORS

 new():
         From Class::Utils::set_params():
                 Unknown parameter '%s'.
         From Tags::HTML::new():
                 Parameter 'css' must be a 'CSS::Struct::Output::*' class.
                 Parameter 'tags' must be a 'Tags::Output::*' class.
         Parameter 'img_link_cb' must be a code.
         Parameter 'img_select_cb' must be a code.
         Parameter 'img_src_cb' must be a code.

 process():
         From Tags::HTML::process():
                 Parameter 'tags' isn't defined.
         Bad data image object.

 process_css():
         From Tags::HTML::process_css():
                 Parameter 'css' isn't defined.

=head1 EXAMPLE

=for comment filename=default_grid.pl

 use strict;
 use warnings;

 use CSS::Struct::Output::Indent;
 use Data::Image;
 use Tags::HTML::Image::Grid;
 use Tags::Output::Indent;

 # Object.
 my $css = CSS::Struct::Output::Indent->new;
 my $tags = Tags::Output::Indent->new;
 my $obj = Tags::HTML::Image::Grid->new(
         'css' => $css,
         'tags' => $tags,
 );

 # Images.
 my $image1 = Data::Image->new(
         'comment' => 'Michal from Czechia',
         'url' => 'https://upload.wikimedia.org/wikipedia/commons/a/a4/Michal_from_Czechia.jpg',
 );
 my $image2 = Data::Image->new(
         'comment' => 'Self photo',
         'url' => 'https://upload.wikimedia.org/wikipedia/commons/7/76/Michal_Josef_%C5%A0pa%C4%8Dek_-_self_photo_3.jpg',
 );

 # Process image grid.
 $obj->process([$image1, $image2]);
 $obj->process_css;

 # Print out.
 print $tags->flush;
 print "\n\n";
 print $css->flush;

 # Output:
 # <div class="image-grid">
 #   <div class="image-grid-inner">
 #     <figure>
 #       <img src=
 #         "https://upload.wikimedia.org/wikipedia/commons/a/a4/Michal_from_Czechia.jpg"
 #         >
 #       </img>
 #       <figcaption>
 #         Michal from Czechia
 #       </figcaption>
 #     </figure>
 #     <figure>
 #       <img src=
 #         "https://upload.wikimedia.org/wikipedia/commons/7/76/Michal_Josef_%C5%A0pa%C4%8Dek_-_self_photo_3.jpg"
 #         >
 #       </img>
 #       <figcaption>
 #         Self photo
 #       </figcaption>
 #     </figure>
 #   </div>
 # </div>
 # 
 # .image-grid {
 # 	display: flex;
 # 	align-items: center;
 # 	justify-content: center;
 # }
 # .image-grid-inner {
 # 	display: grid;
 # 	grid-gap: 1px;
 # 	grid-template-columns: repeat(4, 340px);
 # }
 # .image-grid figure {
 # 	object-fit: cover;
 # 	width: 340px;
 # 	height: 340px;
 # 	position: relative;
 # 	overflow: hidden;
 # 	border: 1px solid white;
 # 	margin: 0;
 # 	padding: 0;
 # }
 # .image-grid img {
 # 	object-fit: cover;
 # 	width: 100%;
 # 	height: 100%;
 # 	vertical-align: middle;
 # }
 # .image-grid figcaption {
 # 	margin: 0;
 # 	padding: 1em;
 # 	position: absolute;
 # 	z-index: 1;
 # 	bottom: 0;
 # 	left: 0;
 # 	width: 100%;
 # 	max-height: 100%;
 # 	overflow: auto;
 # 	box-sizing: border-box;
 # 	transition: transform 0.5s;
 # 	transform: translateY(100%);
 # 	background: rgba(0, 0, 0, 0.7);
 # 	color: rgb(255, 255, 255);
 # }
 # .image-grid figure:hover figcaption {
 # 	transform: translateY(0%);
 # }
 # .image-grid .selected {
 # 	border: 1px solid black;
 # 	border-radius: 0.5em;
 # 	color: black;
 # 	padding: 0.5em;
 # 	position: absolute;
 # 	right: 10px;
 # 	top: 10px;
 # }

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<List::MoreUtils>,
L<Scalar::Util>,
L<Tags::HTML>,
L<Unicode::UTF8>.

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Tags-HTML-Image-Grid>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2022 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.04

=cut
